Improve support for manifest lists and image indexes

nalind opened the pull request:
This patch series adds a `ManifestList` interface, implemented by `Schema2List` and `OCI1Index` types, to the `manifest` package.
It then adds an `instanceDigest` parameter to the `ImageSource.LayerInfosForCopy()`, `ImageDestination.PutManifest()`, and `ImageDestination.PutSignatures()` methods, so that we can store multiple manifests to transports that can support them, and read them from sources which need to update the layer digest list.
It uses those parameters and updates `copy.Image()` to accept a `MultipleImages` boolean option which directs it to copy all instances if the source reference produces an `ImageSource` that represents multiple images.
It then finishes updating the `containers-storage`, `directory`, `docker`, and `oci` transports to be able to read and write multiple images at a time.

nalind commented:
(There's a `PutManifest()` call in skopeo's `cmd/skopeo/layers.go` that we can't update here, so the `make test-skopeo` step will fail in Travis.)

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+	}
+	sort.Strings(hexes)
+	return digest.FromBytes([]byte(strings.Join(hexes, ""))).Hex()
+}
```
This is a guesswork implementation.  If you have a better idea or know of a reference for a better way to compute a stable value, we should replace this.

nalind replied:
Right, that information doesn't seem to live in the image structure, either, which would have been handy.  Changed it to hopefully work more like what you describe.

nalind commented on copy/copy.go:
```diff
@@ -100,6 +102,7 @@ type Options struct {
 	Progress         chan types.ProgressProperties // Reported to when ProgressInterval has arrived for a single artifact+offset.
 	// manifest MIME type of image set by user. "" is default and means use the autodetection to the the manifest MIME type
 	ForceManifestMIMEType string
+	MultipleImages        bool // If set, expect to copy multiple images
```
I keep going back and forth on whether this should be an enumerated value, to be able to request different behavior (for example, "converting" a single image into a multiple image on copy) in the future.

nalind replied:
Makes sense to me.  Changed.

mtrmac commented on types/types.go:
```diff
@@ -180,8 +180,8 @@ type ImageDestination interface {
-	PutManifest(manifest []byte) error
-	PutSignatures(signatures [][]byte) error
+	PutManifest(manifest []byte, instanceDigest *digest.Digest) error
+	PutSignatures(signatures [][]byte, instanceDigest *digest.Digest) error
```
All three updated methods should document `instanceDigest` (and, ideally, this should be propagated all over the copy&pasted comments; that part is non-blocking for me).

nalind replied:
Okay, added.

mtrmac commented on directory/directory_test.go:
```diff
@@ -31,10 +31,14 @@ func TestGetPutManifest(t *testing.T) {
-	err = dest.PutManifest(man)
+	err = dest.PutManifest(man, &md)
+	assert.NoError(t, err)
+	err = dest.PutManifest(man, nil)
```
The two written manifests should be different…

nalind replied:
Ah, right.  Fixed.

mtrmac commented on directory/directory_test.go:
```diff
@@ -47,11 +51,10 @@ func TestGetPutManifest(t *testing.T) {
-	// Non-default instances are not supported
-	md, err := manifest.Digest(man)
-	require.NoError(t, err)
 	_, _, err = src.GetManifest(&md)
```
… to make sure the `GetManifest`/`PutManifest` pairs read/write the correct ones, or, in this case, that the test does not ignore the return value of `GetManifest(&md)` by using `_, _, err`.

nalind replied:
Agreed, fixed.

mtrmac commented on directory/directory_test.go:
```diff
@@ -144,12 +147,19 @@ func TestGetPutSignatures(t *testing.T) {
-	err = dest.PutSignatures(signatures)
+	err = dest.PutSignatures(signatures, nil)
+	assert.NoError(t, err)
+	err = dest.PutSignatures(signatures, &md)
```
Similarly, using two different sets of signatures would be better.

nalind replied:
Agreed, fixed.

mtrmac commented on docker/docker_image_dest.go:
```diff
@@ -213,17 +214,22 @@ func (d *dockerImageDestination) ReapplyBlob(info types.BlobInfo) (types.BlobInf
+		digest = *instanceDigest
+		refTail = instanceDigest.String()
+	} else {
+		d.manifestDigest = digest
```
(Both the `digest` and `refTail` computations can be moved inside this `else`.)

nalind replied:
Makes sense.  Fixed.

mtrmac commented on docker/docker_image_dest.go:
```diff
@@ -297,14 +303,17 @@ func (d *dockerImageDestination) putSignaturesToLookaside(signatures [][]byte) e
+			return errors.Errorf("Unknown manifest digest, can't add signatures")
+		}
+		instanceDigest = &d.manifestDigest
 	}
```
(This repeated block of code could be hoisted up to `PutSignatures`.)

nalind replied:
Doh.  Yeah, fixed that.

mtrmac commented on docker/docker_image_src.go:
```diff
@@ -210,10 +221,11 @@ func (s *dockerImageSource) manifestDigest(ctx context.Context, instanceDigest *
 		}
 	}
-	if err := s.ensureManifestIsLoaded(ctx); err != nil {
+	if err := s.ensureManifestIsLoaded(ctx, instanceDigest); err != nil {
```
Above this, `manifestDigest` starts with `if instanceDigest != nil { return *instanceDigest }`.  And because `cachedManifest` exists only to avoid fetching the manifest second time below that check, AFAICT we only need the single `cachedManifest`/`cachedManifestMIMEType` value for the primary instance.

So, the only necessary change to this file is AFAICT the `LayerInfosForCopy` API update.

nalind replied:
Hmm, I'm not sure how to keep GetManifest() from returning the cached value of the primary instance when `instanceDigest` is not nil otherwise.  Should it just be handing it off to `fetchManifest()` directly for those cases?

mtrmac replied:
Yes, AFAICS. The comment above `ensureManifestIsLoaded` says that the caching exists only for `GetSignatures(_, nil)` to work reliably.

Manifest caching primarily happens in `image.UnparsedImage`, to benefit all transports.

mtrmac commented on openshift/openshift.go:
```diff
@@ -396,17 +396,17 @@ func (d *openshiftImageDestination) ReapplyBlob(info types.BlobInfo) (types.Blob
 	}
 	d.imageStreamImageName = manifestDigest.String()
 
-	return d.docker.PutManifest(m)
```
Above, `d.imageStreamImageName` should only be recorded for the primary instance.

nalind replied:
Oh, right.  That makes sense.  Fixed.

mtrmac replied:
The `d.imageStreamImageName = manifestDigest.String()` assignment is still unconditional AFAICS.

mtrmac commented on openshift/openshift.go:
```diff
@@ -396,17 +396,17 @@ func (d *openshiftImageDestination) ReapplyBlob(info types.BlobInfo) (types.Blob
 
-func (d *openshiftImageDestination) PutSignatures(signatures [][]byte) error {
+func (d *openshiftImageDestination) PutSignatures(signatures [][]byte, instanceDigest *digest.Digest) error {
 	if d.imageStreamImageName == "" {
```
Corresponding to `openshiftImageSource.GetSignatures`, this should use `*instanceDigest` if available and `d.imageStreamImageName` if not.

(Well, I guess. I’m not quite sure anyone has tested multi-arch images with OpenShift to be honest.)

nalind replied:
Right, makes sense.  I'm not sure of that, either.

mtrmac commented on storage/storage_test.go:
```diff
@@ -509,6 +510,341 @@ func TestWriteRead(t *testing.T) {
 	}
 }
 
+func TestWriteReadMulti(t *testing.T) {
```
(Non-blocking: A fair amount of this could be extracted into helpers (e.g. ”verify signatures match expected data”) shared between this and pre-existing tests.

mtrmac commented on storage/storage_image.go:
```diff
@@ -159,35 +162,80 @@ func (s *storageImageSource) getBlobAndLayerID(info types.BlobInfo) (rc io.ReadC
 	return rc, n, layer.ID, err
 }
 
+func (s *storageImageSource) instanceFor(instanceDigest *digest.Digest) string {
```
(Non-blocking: This could be an ordinary function shared between source and destination.)

nalind replied:
Done.

mtrmac commented on storage/storage_image.go:
```diff
@@ -476,14 +541,167 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+			names = append(names, verboseName(name))
+		}
+		if len(oldNames) > 0 {
+			names = append(names, oldNames...)
```
Can this add `name` redundantly even if it is already present?

nalind replied:
In this `names` list, yes, but the storage library deduplicates the lists that are passed to all of its `SetNames()` methods.  I can add logic to also do that here if you prefer.

mtrmac replied:
No need to compute the same thing twice.

Maybe add a comment explaining that `SetNames` does this? OTOH this is no different from any other case where the reader is expected to understand how the called interface works :)

mtrmac replied:
> the reader

i.e. lazy me

nalind commented on storage/storage_image.go:
```diff
@@ -159,35 +162,80 @@ func (s *storageImageSource) getBlobAndLayerID(info types.BlobInfo) (rc io.ReadC
 	}
 	updatedBlobInfos := []types.BlobInfo{}
 	layerID := simg.TopLayer
 	_, manifestType, err := s.GetManifest(nil)
```
Whoops, this should be using `instanceDigest`.

nalind replied:
Fixed that.

nalind commented on copy/copy.go:
```diff
@@ -157,22 +160,33 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
 		}
-		logrus.Debugf("Source is a manifest list; copying (only) instance %s", instanceDigest)
-		unparsedInstance := image.UnparsedInstance(rawSource, &instanceDigest)
+		if !supportsMultipleImages(c.dest) || !options.MultipleImages {
```
Changed it.

nalind commented on copy/copy.go:
```diff
@@ -157,22 +160,33 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
-		unparsedInstance := image.UnparsedInstance(rawSource, &instanceDigest)
+		if !supportsMultipleImages(c.dest) || !options.MultipleImages {
+			// This is a manifest list, and we weren't asked to copy multiple images. Choose a single image and copy it.
+			instanceDigest, err := image.ChooseManifestInstanceFromManifestList(options.SourceCtx, unparsedToplevel)
```
Yeah, I forgot that the version in `image` only knows about one of the list types.  Changed it.

nalind commented on copy/copy.go:
```diff
@@ -183,50 +197,211 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+			Size:      int64(len(updatedManifest)),
+			MediaType: updatedManifestType,
+		}
+		updates = append(updates, update)
```
Changed it.

nalind commented on copy/copy.go:
```diff
@@ -183,50 +197,211 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+
+	// Encode the list again.  If the result is the same as it was before we applied the
+	// updates, then the list wasn't changed, and we can just use the original encoded value.
+	updatedManifestList, err := list.Serialize()
```
Sure.  Changed it.

nalind commented on copy/copy.go:
```diff
@@ -183,50 +197,211 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+	// Determine if we need to convert the manifest to a different format.
+	allowedTypes := c.dest.SupportedManifestMIMETypes()
+	if len(allowedTypes) > 0 {
+		selectedType := allowedTypes[0]
```
Broke it out.

nalind commented on copy/copy.go:
```diff
@@ -183,50 +197,211 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+				if err != nil {
+					return errors.Wrapf(err, "Error encoding updated manifest list %q", string(manifestList))
+				}
+			case "":
```
The broken-out version of this can treat this as an error because you're right, it shouldn't happen.

nalind commented on copy/copy.go:
```diff
@@ -183,50 +197,211 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+	}
+
+	// If we can't use the original value, flag an error.
+	if !bytes.Equal(updatedManifestList, originalManifestList) {
```
Updated this to just reference a boolean that we set when we know we're making or have made changes of significance.

nalind commented on copy/copy.go:
```diff
@@ -183,50 +197,211 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
 	// Please keep this policy check BEFORE reading any other information about the image.
 	// (the multiImage check above only matches the MIME type, which we have received anyway.
-	// Actual parsing of anything should be deferred.)
+	// Actual parsing of anything should be deferred.
```
Ah, right.  Fixed.

nalind commented on copy/copy.go:
```diff
@@ -183,50 +197,211 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
 	} else {
 		c.Printf("Getting image source signatures\n")
-		s, err := src.Signatures(context.TODO())
+		s, err := c.rawSource.GetSignatures(context.TODO(), sourceInstance)
```
I must have misread how `UnparsedImage.Signatures()` was doing things -- for some reason I kept thinking it only returned the primary manifest's signatures.  Reverting this bit.

mtrmac replied:
Please drop the `sourceInstance` parameter to `copyOneImage` as well.

nalind commented on copy/copy.go:
```diff
@@ -457,7 +634,11 @@ func (ic *imageCopier) copyUpdatedConfigAndManifest() ([]byte, error) {
+		tmpDigest := digest.FromBytes(manifest)
+		instanceDigest = &tmpDigest
+	}
+	if err := ic.c.dest.PutManifest(manifest, instanceDigest); err != nil {
```
I'm not sure I follow your meaning.  Either `instanceDigest` was `nil`, and we didn't change it, or it wasn't, and we might have.

mtrmac replied:
I’m sorry, I’m not making sense to myself either.  I guess I misread `if instanceDigest != nil` as `if instanceDigest == nil`.

nalind commented on copy/copy.go:
```diff
@@ -457,7 +634,11 @@ func (ic *imageCopier) copyUpdatedConfigAndManifest() ([]byte, error) {
+		tmpDigest := digest.FromBytes(manifest)
+		instanceDigest = &tmpDigest
+	}
+	if err := ic.c.dest.PutManifest(manifest, instanceDigest); err != nil {
```
`manifest` here is being read from `pendingImage`, which should have had updates applied to it at this point.  Being able to call `manifest.Digest()` (which, yeah, we need to be consistent about that) means I need to rename the variable, though, so that's changed.  Using `MatchesDigest()` just forces us to compute the digest a second time if they don't match, though, so I don't see much of a gain from making that a second condition.

mtrmac replied:
> Using MatchesDigest() just forces us to compute the digest a second time if they don't match, though, so I don't see much of a gain from making that a second condition.

In _theory_ a `manifest.MatchesDigest()` check would allow preserving a digest using a non-default algorithm. In _practice_, yeah, `manifest.MatchesDigest` is just a `manifest.Digest` + a comparison, there isn’t currently a difference. Still, doing the theoretically right thing here would allow us to extend `manifest.MatchesDigest` in the future.

We can compensate for the repeated manifest digest computation by returning the computed manifest digest from this function and `copyOneImage`, and dropping
```go
		// Record the result of a possible conversion here.
		md, err := manifest.Digest(updatedManifest)
```
in `copyMultipleImages` ;-)

nalind commented on image/oci_index.go:
```diff
@@ -0,0 +1,57 @@
+
+// chooseDigestFromImageIndex parses blob as an OCI1 image index,
+// and returns the digest of the image appropriate for the current environment.
+func chooseDigestFromImageIndex(ctx *types.SystemContext, blob []byte) (digest.Digest, error) {
```
Changed it.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,191 @@
+
+// Schema2PlatformSpec describes the platform which a particular manifest is
+// specialized for.
+type Schema2PlatformSpec struct {
```
Changed it.

mtrmac replied:
This regressed it seems.

nalind replied:
Argh.  The code changes stuck, but the old definitions were still there.  Removing them.

nalind replied:
Should be gone now.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,191 @@
+
+// ChooseDigest parses blob as a schema2 manifest list, and returns the digest
+// of the image which is appropriate for the current environment.
+func (list *Schema2List) ChooseDigest(ctx *types.SystemContext) (digest.Digest, error) {
```
Changed it.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,191 @@
+
+// ToSchema2List() returns the list encoded as a Schema2 list.
+func (list *Schema2List) ToSchema2List() (*Schema2List, error) {
+	return Schema2ListClone(list), nil
```
Maybe not, but it's less confusing for me that way.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,191 @@
+
+// Schema2ListFromManifest creates a Schema2 manifest list instance from marshalled
+// JSON, presumably generated by encoding a Schema2 manifest list.
+func Schema2ListFromManifest(manifest []byte) (*Schema2List, error) {
```
These names were an attempt to be consistent with the `...FromManifest()` functions that return specific single-image manifest types, rather than with `FromBlob()`, which returns the generic interface.  If you're sure you prefer it the other way, though, I'll change it.

mtrmac replied:
Oops, the two probably should have been consistent from the start. My mistake.

You’re right, …`FromManifest` makes more sense given the other ones now.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,191 @@
+func Schema2ListFromManifest(manifest []byte) (*Schema2List, error) {
+	list := Schema2List{
+		SchemaVersion: 2,
+		MediaType:     DockerV2ListMediaType,
```
For some reason I must have thought that they were `omitEmpty`.  Removed.  Do we want to do any checking after the decode to ensure that they were set to the expected values?

mtrmac replied:
Sanity checking here would be nice, but isn’t strictly necessary. From the pre-existing manifest types, only schema1 does some checking.

(In many cases, we call `manifest.GuessMIMEType`, which parses enough of the manifest to make essentially the same sanity check. So I’m not sure how much checking in here would matter in practice. In principle, of course, relying on `GuessMIMEType` is very unclean—ideally we would want to get rid of it if at all possible.)

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+
+	// ChooseDigest selects which manifest is most appropriate for the platform described by the
+	// SystemContext, or for the current platform if the SystemContext doesn't specify any details.
+	ChooseDigest(ctx *types.SystemContext) (digest.Digest, error)
```
Changed it.

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+	ToOCI1Index() (*OCI1Index, error)
+
+	// ToSchema2List() returns the list rebuilt as a Schema2 list, converting it if necessary.
+	ToSchema2List() (*Schema2List, error)
```
Simple enough.  Added it.

mtrmac replied:
Could `copyMultipleImages` call `ConvertToMIMEType` now?

mtrmac replied:
Could this go one step further, then, and having only the `ConvertToMIMEType` method public, and having each individual implementation its own dispatch into private conversion implementations, without everything indirecting through `manifest.ConvertListToMIMEType?

Having one `switch` per implementation should not be that much more code, and it would allow keeping the `List` interface type-agnostic (apart from hard-coding the concept of a manifest list and assumptions about what that means, sure).

nalind replied:
Looks reasonable.  Changing it.

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+	return dup
+}
+
+// dupStringStringMape returns a deep copy of a map[string]string, always returning
```
Doh.  Thanks for spotting that.

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+// an initialized map, even if there are no contents to copy into it.
+func dupStringStringMap(m map[string]string) map[string]string {
+	if len(m) == 0 {
+		return nil
```
Yeah, that had to be adjusted to make some new tests pass.  I've corrected the comment.

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+	case imgspecv1.MediaTypeImageManifest:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
+	case DockerV2Schema2MediaType:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
```
There isn't AFAIK.  This was mimicking the behavior in `FromBlob()`, but since we're not customizing the error message based on the type, you're right, there's no need for that.  Collapsed them.

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+	default:
+		return nil, fmt.Errorf("Unimplemented manifest MIME type %s (normalized as %s)", manifestMIMEType, normalized)
+	}
+	return nil, fmt.Errorf("Unimplemented manifest MIME type %s", manifestMIMEType)
```
Yup.  Removed the `default` case.

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+
+// ChooseManifestInstanceFromManifestList returns a digest of a manifest appropriate
+// for the current system from the manifest available from src.
+func ChooseManifestInstanceFromManifestList(ctx *types.SystemContext, manifests ManifestList) (digest.Digest, error) {
```
Makes sense.  Dropped it.

mtrmac replied:
`image.ChooseManifestInstanceFromManifestList` exists again.

nalind replied:
Huh.  Dropping it again.

nalind replied:
Okay, it should be gone now.

nalind commented on manifest/list.go:
```diff
@@ -0,0 +1,119 @@
+	if mt != DockerV2ListMediaType && mt != imgspecv1.MediaTypeImageIndex {
+		return "", fmt.Errorf("Internal error: Trying to select an image from a non-manifest-list manifest type %s", mt)
+	}
+	return manifests.ChooseDigest(ctx)
```
Its only user at this point was `image.Copy()`, so it's gone now.

nalind commented on manifest/list_test.go:
```diff
@@ -0,0 +1,93 @@
+
+		m, err := ListFromBlob(manifest, c.mimeType)
+		require.NoError(t, err, "manifest list %q  should parse as list types", c.path)
+		assert.True(t, c.checkType(m), "manifest %q is not of the expected implementation type", c.path)
```
`MIMEType()` seems less messy to use, changed it.  Likewise, now that we have a `Clone()` method, switched to using that.

nalind commented on manifest/manifest.go:
```diff
@@ -180,7 +181,7 @@ func AddDummyV2S1Signature(manifest []byte) ([]byte, error) {
 // MIMETypeIsMultiImage returns true if mimeType is a list of images
 func MIMETypeIsMultiImage(mimeType string) bool {
-	return mimeType == DockerV2ListMediaType
+	return mimeType == DockerV2ListMediaType || mimeType == imgspecv1.MediaTypeImageIndex
```
Right, I should have caught that.  Added.

nalind commented on manifest/manifest.go:
```diff
@@ -222,9 +224,12 @@ func FromBlob(manblob []byte, mt string) (Manifest, error) {
-		return nil, fmt.Errorf("Unimplemented manifest MIME type %s", mt)
+		return nil, fmt.Errorf("Unimplemented manifest MIME type %s (normalized as %s)", mt, nmt)
 	}
+	return nil, fmt.Errorf("Unimplemented manifest MIME type %s", mt)
```
It should be.  Removed the `default` clause to let us log the normalized value.

mtrmac replied:
Please preserve the `Note that this may not be reachable, NormalizedMIMEType has a default for unknown values.` comment.

mtrmac replied:
Please preserve the `Note that this may not be reachable, NormalizedMIMEType has a default for unknown values.` comment.

nalind replied:
Doh.  Thanks for reminding me, as I missed it.  Adding it back after the end of the `switch` statement.

nalind replied:
This should be there now.

nalind commented on manifest/oci_index.go:
```diff
@@ -0,0 +1,194 @@
+
+// ToOCI1Index() returns the index encoded as an OCI1 index.
+func (index *OCI1Index) ToOCI1Index() (*OCI1Index, error) {
+	return OCI1IndexClone(index), nil
```
It makes it easier to predict how modifying the returned value will affect the original when it's called through the `ManifestList` interface, so I like it for that reason.  If you have a strong preference for just returning `index` here, though, I'll change it.

mtrmac replied:
No, cloning is fine of course.

nalind commented on oci/archive/oci_src.go:
```diff
@@ -90,6 +90,6 @@ func (s *ociArchiveImageSource) GetSignatures(ctx context.Context, instanceDiges
 // LayerInfosForCopy() returns updated layer info that should be used when reading, in preference to values in the manifest, if specified.
-func (s *ociArchiveImageSource) LayerInfosForCopy() []types.BlobInfo {
+func (s *ociArchiveImageSource) LayerInfosForCopy(*digest.Digest) []types.BlobInfo {
 	return nil
```
Makes sense.  Changed it.

nalind commented on oci/layout/oci_dest.go:
```diff
@@ -191,20 +192,20 @@ func (d *ociImageDestination) ReapplyBlob(info types.BlobInfo) (types.BlobInfo,
 // If the destination is in principle available, refuses this manifest type (e.g. it does not recognize the schema),
 // but may accept a different manifest type, the returned error must be an ManifestTypeRejectedError.
-func (d *ociImageDestination) PutManifest(m []byte) error {
+func (d *ociImageDestination) PutManifest(m []byte, instanceDigest *digest.Digest) error {
```
No, you've got it right.  This preserves enough information that, if we're copying an index of images into a layout, we'll be able to read everything back by using the tag ("image") name, and also be able to grab any one instance by specifying its digest, for cases where we're copying an image out, and the destination for that copy can only accept one image.

mtrmac replied:
Good point, users will expect to be able to use the manifest digests.

Still, it feels risky; e.g. the annotations or platform selectors may easily get out of sync between the duplicate manifest descriptors. And once we start generating images like this I’m not certain we will be able to stop without hurting consumers who start to rely on it.

OTOH this layout is fairly simple to integrate into the `ImageSource`/`ImageDestination` interfaces, avoiding the duplicates would be much messier at a first glance.

@cyphar WDYT?  (Context is representing multi-arch images in an OCI repo, similar to schema2 manifest lists.)

nalind commented on oci/layout/oci_dest.go:
```diff
@@ -191,20 +192,20 @@ func (d *ociImageDestination) ReapplyBlob(info types.BlobInfo) (types.BlobInfo,
 // but may accept a different manifest type, the returned error must be an ManifestTypeRejectedError.
-func (d *ociImageDestination) PutManifest(m []byte) error {
+func (d *ociImageDestination) PutManifest(m []byte, instanceDigest *digest.Digest) error {
 	digest, err := manifest.Digest(m)
```
Ah, right.  Fixed.

nalind commented on oci/layout/oci_dest.go:
```diff
@@ -218,33 +219,108 @@ func (d *ociImageDestination) PutManifest(m []byte) error {
+	case imgspecv1.MediaTypeImageManifest:
+		// Continue filling out the descriptor.
+		desc.MediaType = imgspecv1.MediaTypeImageManifest
+		// Try to figure out what the Platform field should contain.
```
Now that I think about it more, `ChooseInstance()` doesn't care about platform information in the layout's internal index, since it's searching the primary manifest (which is an index), which has the information.  So yeah, dropped that.

nalind commented on oci/layout/oci_dest.go:
```diff
@@ -218,33 +219,108 @@ func (d *ociImageDestination) PutManifest(m []byte) error {
+				return
+			}
+			// Replace the other fields that might have changed, even if the content didn't.
+			d.index.Manifests[i].MediaType = desc.MediaType
```
Ah, right, that value is always supplied by `PutManifest()`, so it won't change.  Dropped that.

nalind commented on oci/layout/oci_src.go:
```diff
@@ -71,13 +76,15 @@ func (s *ociImageSource) GetManifest(instanceDigest *digest.Digest) ([]byte, str
-		// and see the MIME type there.
-		mimeType = imgspecv1.MediaTypeImageManifest
+		for _, md := range s.index.Manifests {
+			if md.Digest == dig {
```
I don't think I understand your meaning here.  Do you mean that we should be building an index mapping tag names to digests when we load the index?

mtrmac replied:
This was just pointing out another place which expects the layout described in the “What is the overall layout is this designed to create?” comment.

{Because,As long as} the layout with duplicate descriptors for instance manifests is intentional, this code makes sense.

nalind commented on oci/layout/oci_transport.go:
```diff
@@ -186,9 +186,6 @@ func (ref ociReference) getManifestDescriptor() (imgspecv1.Descriptor, error) {
 	} else {
 		// if image specified, look through all manifests for a match
 		for _, md := range index.Manifests {
```
I don't think I understand what you mean here.  If you're asking if we need to be able to read an image index, when it's the only manifest, by specifying an empty image name, then I think so, since we already supported reading single-image manifests that way.

mtrmac replied:
Yes, we support reading single-image manifests with no name annotation when it’s the only manifest in the repo.

---

I was thinking about cases like this:

Top-level index:
- unnamed descriptor for x86_64
- unnamed descriptor for arm64

(i.e. the top-level index is “the manifest list”; with no nested index entry in the top-level index)

or possibly

Top-level index:
- unnamed descriptor for a manifest list
  - unnamed descriptor for x86_64
  - unnamed descriptor for arm64
- unnamed descriptor for x86_64
- unnamed descriptor for arm64

(i.e. the output generated by `ociImageDestination` now, except with no name annotation).

---

Both cases would currently fail the `if len(index.Manifests) == 1 {` check.

Maybe it’s just not worth worrying about now.

nalind commented on storage/storage_image.go:
```diff
@@ -476,14 +541,167 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+	for layerDigest := range s.allLayerSums {
+		delete(dataBlobs, layerDigest)
+	}
+	for blob := range dataBlobs {
```
Yeah, without making the assumption that the only non-layer in an image is the config blob (the data formats suggest that's safe, but the ImageDestination interface doesn't seem to guarantee that), or making `PutBlob()` and friends take an `instanceDigest`, I couldn't be sure that we wouldn't end up not being able to find a blob if we subsequently tried to read a specific image and its layers and blobs by itself.  If you're telling me that's safe to depend on, I can clean that up.

mtrmac replied:
Ah, you’re right; the `ImageDestination` gets all blobs for all instances in a single `NewImageDestination`…`Commit` sequence.

It seems `containers-storage` is currently the only transport where this matters, so it’s mostly up to you what the interface should look like.

- Maybe this is good enough
- Maybe all `PutBlob`s should get an `instanceDigest` (and we would somehow use `ReapplyBlob` to share blobs across architectures?)
- https://github.com/containers/image/pull/370 adds an `isConfig` boolean parameter to `PutBlob`. Maybe, instead, we should have an instance-independent `PutLayer`, and a `PutConfig` which specifies an image digest
- Something else?

nalind replied:
We merged #370 in the meantime, and that seems to be working well enough here.

nalind commented on storage/storage_image.go:
```diff
@@ -476,14 +541,167 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+			return errors.Wrapf(err, "error copying non-layer blob %q to image", blob)
+		}
+		if err := s.imageRef.transport.store.SetImageBigData(imageID, blob.String(), v); err != nil {
+			if _, err2 := s.imageRef.transport.store.DeleteImage(imageID, true); err2 != nil {
```
At that point the image is in a partially-broken state, and deleting it is probably friendlier in terms of not leaving trash behind if the whole thing isn't reattempted.  Moved the `DeleteImage()` call to the callers.

nalind commented on storage/storage_image.go:
```diff
@@ -476,14 +541,167 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+	if name := s.imageRef.DockerReference(); len(oldNames) > 0 || name != nil {
+		var err error
+		if instanceDigest != nil {
+			name, err = reference.WithDigest(reference.TrimNamed(name), *instanceDigest)
```
Attempting to trim `name` would have created problems.  I've added a check around this, so that we don't try to add the digested reference.

nalind commented on storage/storage_image.go:
```diff
@@ -476,14 +541,167 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+			names = append(names, oldNames...)
+		}
+		if err = s.imageRef.transport.store.SetNames(imageID, names); err != nil {
+			if _, err2 := s.imageRef.transport.store.DeleteImage(imageID, true); err2 != nil {
```
Changed, as above.

nalind commented on storage/storage_image.go:
```diff
@@ -476,14 +541,167 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+		options.Digest = manifestDigest
+	}
+	// Create the image record, pointing to no layers.
+	intendedID := s.imageRef.id
```
Absolutely.  Refactored them.

TomSweeneyRedHat commented on copy/copy.go:
```diff
@@ -151,65 +170,257 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+		}
+	}
+	// ... but if we're forcing it, we prefer the forced value a bit more.
+	if forcedManifestListMIMEType != "" {
```
can you do the for loop from 251-255 as part of an else for this if statement?

nalind commented:
The linter's complaining about stuttering with the `ManifestList` and `ManifestListUpdate` types.  Any preferences over renaming them to `List` and `ListUpdate`?

TomSweeneyRedHat commented on storage/storage_image.go:
```diff
@@ -476,14 +550,175 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+	// we just need to screen out the ones that are actually layers to get the list of non-layers.
+	if err = s.commitDataBlobs(img.ID); err != nil {
+		if _, err2 := s.imageRef.transport.store.DeleteImage(img.ID, true); err2 != nil {
+			logrus.Debugf("error deleting incomplete image %q: %v", img.ID, err2)
```
On the fence here, but if you've an error deleting an incomplete image, should it just go to debug  or should you wrap up err2 also and return with the return below?

TomSweeneyRedHat replied:
Ditto with the next few delete with a debug on error.

mtrmac replied:
`copy.Image` does that in a few (very few) cases, see `if err := dest.Close(); err != nil {`. All too often we do `defer something.Close()` and ignore the possible failure.

Either way it’s problematic. On one hand, yes, throwing away failure information may make diagnosing failures difficult. On the other, the `copy.Image` attempt to parenthesize the secondary failure has already confused at least one user, who focused on the secondary failure instead of the primary one.

TomSweeneyRedHat commented on storage/storage_image.go:
```diff
@@ -476,14 +550,175 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+			logrus.Debugf("error deleting incomplete image %q: %v", img.ID, err2)
+		}
+		logrus.Debugf("error saving manifest for image %q: %v", img.ID, err)
+		return err
```
The errors prior to this you wrapped with a little verbiage.  This one and below you did a debug then straight error.  I'd vote continuing on with the error wrap using the veribage from the debugs, but your call.

nalind replied:
Changed them as you described.

mtrmac commented on copy/copy.go:
```diff
@@ -151,65 +170,226 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
-		// This is a manifest list. Choose a single image and copy it.
-		// FIXME: Copy to destinations which support manifest lists, one image at a time.
-		instanceDigest, err := image.ChooseManifestInstanceFromManifestList(options.SourceCtx, unparsedToplevel)
+		// If we were asked to copy multiple images and can't, that's an error.
```
This comment seems to belong at the top of the `else` branch below.

mtrmac commented on copy/copy.go:
```diff
@@ -151,65 +170,226 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+	}
+
+	// Determine if we need to convert the manifest to a different format.
+	selectedType, _, err := c.determineManifestListConversion(manifestType, c.dest.SupportedManifestMIMETypes(), "")
```
I’m sorry, by “similar to `determineManifestConversion`” I didn’t mean necessary literally equal; including the logic for the `forcedManifestListMIMEType` parameter, which is actually hard-coded to `""` here, and the `otherTypes` return value, which is ignored here, is not useful AFAICT, only confusing dead code.

mtrmac replied:
The returned `otherTypes` list is still ignored in the only caller.

nalind replied:
`determineListConversion` no longer returns that value.

mtrmac commented on manifest/list_test.go:
```diff
@@ -0,0 +1,151 @@
+	return nil
+}
+
+func cloneSchema2List(i interface{}) ManifestList {
```
The four helper methods above are no longer used.

mtrmac replied:
Still outstanding.

nalind replied:
Okay, they should be gone now.

mtrmac commented:
> The linter's complaining about stuttering with the `ManifestList` and `ManifestListUpdate` types.  Any preferences over renaming them to `List` and `ListUpdate`?

`manifest.List` works for me. Definitely better than `manifest.MultiArchImage` :)

mtrmac commented on copy/copy.go:
```diff
@@ -151,65 +170,226 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+	if listIsModified {
+		canModifyManifest := (len(sigs) == 0)
+		if !canModifyManifest {
+			return errors.Errorf("Internal error: copyMultipleImages() needs to use an updated manifest but that was known to be forbidden")
```
The `Internal error` wording in the old code was based on the expectation that `if !canModifyManifest`, the code doing the modification should have already noticed and failed with a more descriptive error message (e.g. refused to convert the manifest). This is not happening here.

Still, the simplest solution may be to drop any signature support for manifest lists, as discussed elsewhere.

mtrmac replied:
Still outstanding.

nalind replied:
Okay, moved this check higher up.

mtrmac commented on manifest/list.go:
```diff
@@ -0,0 +1,133 @@
+	case DockerV2Schema1MediaType, DockerV2Schema1SignedMediaType, imgspecv1.MediaTypeImageManifest, DockerV2Schema2MediaType:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
+	}
+	return nil, fmt.Errorf("Unimplemented manifest MIME type %s (normalized as %s)", manifestMIMEType, normalized)
```
(Maybe this should say “manifest list MIME type” to allow distinguishing this failure from `manifest.FromBlob` failing.)

mtrmac commented on manifest/list_test.go:
```diff
@@ -0,0 +1,151 @@
+			man, err := ioutil.ReadFile(filepath.Join("..", "image", "fixtures", manifestList.listFile))
+			require.NoError(t, err)
+			manifestList.rawManifest = man
+		}
```
AFAICT `len(manifestList.listFile) > 0` is always true, and `manifestList.rawManifest` could be a local variable in this block instead of a `struct` member.

mtrmac commented on oci/layout/oci_dest_test.go:
```diff
@@ -119,6 +145,28 @@ func putTestManifest(t *testing.T, ociRef ociReference, tmpDir string) {
 	assert.Contains(t, paths, filepath.Join(tmpDir, "blobs", "sha256", digest), "The OCI directory does not contain the new manifest data")
 }
+
+func putTestIndex(t *testing.T, ociRef ociReference, tmpDir string) {
```
This is unused AFAICS.

mtrmac commented on storage/storage_image.go:
```diff
@@ -476,14 +550,175 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+// supplied, a tag or digest from the reference is replaced with the digest.
+func (s *storageImageDestination) commitName(imageID string, oldNames []string, instanceDigest *digest.Digest) error {
+	if name := s.imageRef.DockerReference(); len(oldNames) > 0 || name != nil {
+		var err error
```
(This declaration seems unnecessary. Doesn’t hurt either.)

mtrmac commented on storage/storage_image.go:
```diff
@@ -476,14 +550,175 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
-	man, err := manifest.FromBlob(s.manifest, manifest.GuessMIMEType(s.manifest))
+	man, err := manifest.ListFromBlob(m, s.manifestTypes[instance])
+	if err != nil {
+		return errors.Wrapf(err, "Internal error: storageImageDestination.Commit() called without PutManifest()")
```
This should probably say something about parsing the manifest list.

rhatdan commented:
@nalind is this something you are still interested in or should we close this?

mtrmac commented:
> @nalind is this something you are still interested in or should we close this?

The _feature_ is definitely useful and eventually we would like to have it; this PR is about 90% there, so it seems more useful to keep it open and noticed once in a few months than to close it and have it forgotten forever.

bowlofeggs commented:
I would also really like OCI and manifest list support in Skopeo for https://github.com/fedora-infra/bodhi

runcom commented:
@nalind could you rebase this? :trollface:

rhatdan commented:
@nalind Either rebase or close?

fabiand commented:
Rebase!

rhatdan commented:
@nalind @vrothberg @mtrmac Any chance we can get movement on this?  If @nalind Does not have the time, could someone else take it over?

vrothberg commented:
> @nalind @vrothberg @mtrmac Any chance we can get movement on this? If @nalind Does not have the time, could someone else take it over?

I have too many things in the pipeline. If that's a rather urgent feature, I suggest we plan for it.

mtrmac commented on image/oci_index.go:
```diff
@@ -0,0 +1,34 @@
+		return nil, errors.Wrap(err, "Error computing manifest digest")
+	}
+	if !matches {
+		return nil, errors.Errorf("Manifest image does not match selected manifest digest %s", targetManifestDigest)
```
s/`Manifest image`/`Image manifest/`

nalind replied:
Dang it, thought I'd fixed that.  Fixing it.

mtrmac commented:
> Squashed, rebased, and changed a few things.

Thanks!

> The big change is in the way we're handling lists of images for the `storage` backend, in that it no longer claims to support manifest lists. But `copy.Image()` now passes the top-level `UnparsedImage` in through the `context.Context` (probably not a good practice), and `storageImageDestination` stores its manifest as a second manifest in the local copy of the image.

I’ve been thinking about this… it seems cleanest to me to add something like `MultipleImagesOption.SparseManifestList` (maybe not a part of the public API, but at least internally) which causes the copy to copy only the current image + unchanged manifest list. And then the transports (sources+destinations) would indicate whether they require/contain sparse images.

Notably this would mean that repo@digest references referencing the manifest list can continue to work, without having to break digest verification.

OTOH we would have to nail down the exact semantics, for sources (does `copy.Image` explicitly try to read only that one image? Does it try all referenced images, and just silently ignore failed architecture?) and destinations (what happens a copy from a sparse source to a non-sparse that requires all instances to exist?).

Not sure yet, only starting to read the updated PR now.

mtrmac commented on image/docker_list.go:
```diff
@@ -58,7 +58,11 @@ func chooseDigestFromManifestList(sys *types.SystemContext, blob []byte) (digest
 }
 
 func manifestSchema2FromManifestList(ctx context.Context, sys *types.SystemContext, src types.ImageSource, manblob []byte) (genericManifest, error) {
-	targetManifestDigest, err := chooseDigestFromManifestList(sys, manblob)
```
`chooseDigestFromManifestList` should be removed in favor of calling `manifest.Schema2List.ChooseInstance` everywhere.

nalind replied:
Dropping the function.

nalind commented:
> I’ve been thinking about this… it seems cleanest to me to add something like `MultipleImagesOption.SparseManifestList` (maybe not a part of the public API, but at least internally) which causes the copy to copy only the current image + unchanged manifest list. And then the transports (sources+destinations) would indicate whether they require/contain sparse images.

This would make it a little easier to avoid problems that would crop up if we pulled down each of the images in a list individually using their repo@digest references, and also the list's repo@digest reference, which would end up attaching the manifest list to the os/arch-specific image.  Even with all of the pieces present in local storage, a copy using `CopyAllImages` won't succeed because `GetManifest()` only looks at one containers/storage Image record, and the other arches are stored as other containers/storage Image records.

> Notably this would mean that repo@digest references referencing the manifest list can continue to work, without having to break digest verification.

That should work in the current iteration - when we look up an image using a repo@digest reference, we're careful to return a matching manifest if the image has one.

If we somehow pull using the list's repo@digest reference twice, _using different `ArchitectureChoice` and/or `OSChoice` values each time_, though, we produce two containers/storage Image records which contain the manifest list, and I probably need to add logic to `storageReference.resolveImage()` to make sure it selects among possibly-more-than-one result from `s.transport.store.ImagesByDigest()`.

> OTOH we would have to nail down the exact semantics, for sources (does `copy.Image` explicitly try to read only that one image? Does it try all referenced images, and just silently ignore failed architecture?) and destinations (what happens a copy from a sparse source to a non-sparse that requires all instances to exist?).

In its current form, attempting to copy from a containers/storage Image that includes a manifest list, using the repo@digest reference for the list and with `CopyAllImages`, will fail.  A dedicated `manifest push` subcommand could fake up enough of an `ImageSource` to stitch together blobs and manifests from some combination of local and remote images, but we could try to handle the local-only case in `storageImageSource`.  I'm not sure I'd want to tackle that in this PR, though.

nalind commented:
> If we somehow pull using the list's repo@digest reference twice, _using different `ArchitectureChoice` and/or `OSChoice` values each time_, though, we produce two containers/storage Image records which contain the manifest list, and I probably need to add logic to `storageReference.resolveImage()` to make sure it selects among possibly-more-than-one result from `s.transport.store.ImagesByDigest()`.

Added logic for that.

displague commented on Makefile:
```diff
@@ -78,6 +78,8 @@ test-skopeo:
 		rm -rf $${vendor_path} && cp -r . $${vendor_path} && rm -rf $${vendor_path}/vendor && \
 		cd $${skopeo_path} && \
+		echo == Patching skopeo, badly - DO NOT MERGE Makefile changes == && \
+		sed -i -e "s:ctx, manifest:ctx, manifest, nil:" cmd/skopeo/layers.go && \
```
`sed -i` won't build on OSX, it will take `-e` as the filename extension as the back copy to create.

nalind replied:
Not to worry.  This commit isn't something we want to keep, but in the meantime it lets us run the skopeo parts of the tests in CI.

nalind commented:
Hmm, I'm running into a problem where `CopyOnlyCurrentRuntimeImage` and `CopyAllImages` aren't flexible enough, as neither lets me copy _only_ a manifest list from one location to another.

I'm considering adding another enumeration value (`CopyOnlyInstances`) that lets the caller specify which instances to copy by listing them in a new field in the `copy.Options` structure (`Instances`, ignored unless `options.MultipleImages == CopyOnlyInstances`).

Size and media type info would make building the partial-update to pass to `list.UpdateInstances()` easier, and adding an `Instance()` method that returns a `manifest.ListUpdate` for a given instance digest would make getting at that easy.

nalind commented:
Added some context to a few errors.  Building manifest lists and storing them in image records in containers/storage trips up some tools because `NewImage()` fails to read the target platform's instance's manifest (which, since we're building a manifest list, isn't there), and they don't handle that well.

nalind commented:
Fixed the `skip` logic, adjusted the debugging message that we print before calling `copyMultipleImages()` to indicate whether we're copying all images, or just some of them, and cherry-picking #711 to avoid a test failure due to using certificates which are now past their not-valid-after dates.

nalind commented:
Rebased, dropping our cherry pick of #711, since it's been merged.

nalind commented:
Rebased, and adjusted tests in the skopeo branch that we're using to expect different error messages we get since https://github.com/containers/storage/pull/436.

nalind commented:
Rebased.

nalind commented:
Rebased.

nalind commented:
Tweaked the description for `CopyOnlyInstances`.

mrunalp commented:
@vrothberg ptal

vrothberg commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,212 @@
+package manifest
```
Nit: rename the file to `docker_schema2_list.go` to be consistent with other files in the package.

nalind replied:
Renamed it.

vrothberg commented on manifest/oci_index.go:
```diff
@@ -0,0 +1,217 @@
+	case imgspecv1.MediaTypeImageIndex:
+		return index.Clone(), nil
+	case DockerV2Schema1MediaType, DockerV2Schema1SignedMediaType, imgspecv1.MediaTypeImageManifest, DockerV2Schema2MediaType:
+		return nil, fmt.Errorf("Can not convert image index to MIME type %q, which is not a list type", manifestMIMEType)
```
Same as for schema2: I think we can use a default case and preserve the error message.

nalind replied:
Changed it, as for `Schema2List.ConvertToMIMEType()`.

vrothberg commented on storage/storage_image.go:
```diff
@@ -292,25 +310,34 @@ func buildLayerInfosForCopy(manifestInfos []manifest.LayerInfo, physicalInfos []
+		key = signatureBigDataKey(*instanceDigest)
+		instance = instanceDigest.Encoded()
+	} else {
+	}
```
The empty `else` branch looks suspicious.

nalind replied:
Yeah, that doesn't need to be there.  Removed it.

vrothberg commented on copy/copy.go:
```diff
@@ -205,80 +238,279 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
 
+	// Give Commit() a pointer to the unparsed source image in its original form.  This may not exactly fit the intended
+	// use for WithValue(), though, so we may need to switch to passing it directly to dest.Commit() and c.copyOneImage().
+	ctx = context.WithValue(ctx, internal.CopyUnparsedTopImageKey, unparsedToplevel)
```
A few hours later and having read more code changes ... we should really add that to the Commit API. We're changing other parts of ImageDestination. Feel free to ignore the `CommitOptions` comment above and add another parameter.

nalind replied:
Okay, added the top level unparsed image as an `UnparsedImage` parameter for `ImageDestination.Commit`.  This moves the point where we retrieve the top-level manifest in the storage ImageDestination later, to `Commit()`-time, so verification of that digest, and of the digest of the image that we're copying has to move to that point as well.

vrothberg commented on storage/storage_reference.go:
```diff
@@ -51,9 +54,58 @@ func imageMatchesRepo(image *storage.Image, ref reference.Named) bool {
 	return false
 }
 
+func imageMatchesSystemContext(store storage.Store, img *storage.Image, manifestDigest digest.Digest, sys *types.SystemContext) bool {
```
Please add a (verbose) comment.

nalind replied:
Added.

nalind commented on copy/copy.go:
```diff
@@ -110,6 +112,32 @@ type imageCopier struct {
+	// empty), along with the list itself.  If the target reference can
+	// only accept one image (i.e., it cannot accept lists), an error
+	// should be returned.
+	CopyOnlyInstances
```
Sure, I'm not attached to the names.  Changed them.

nalind commented on copy/copy.go:
```diff
@@ -110,6 +112,32 @@ type imageCopier struct {
+
+// MultipleImagesOption is either CopyOnlyCurrentRuntimeImage or CopyAllImages, to control
+// whether copy.Image() copies only an image which matches the current runtime environment, or
+// all images which match the supplied reference.
```
Sure, I'm not attached to the names.  Changed it.

nalind commented on copy/copy.go:
```diff
@@ -121,12 +149,14 @@ type Options struct {
 	Progress         chan types.ProgressProperties // Reported to when ProgressInterval has arrived for a single artifact+offset.
 	// manifest MIME type of image set by user. "" is default and means use the autodetection to the the manifest MIME type
 	ForceManifestMIMEType string
+	MultipleImages        MultipleImagesOption // set to either CopyOnlyCurrentRuntimeImage, CopyAllImages, or CopyOnlyInstances
```
Updated.

nalind commented on copy/copy.go:
```diff
@@ -136,6 +166,9 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
 
+	if options.MultipleImages != CopyAllImages && options.MultipleImages != CopyOnlyCurrentRuntimeImage && options.MultipleImages != CopyOnlyInstances {
+		return nil, errors.Errorf("Invalid value for options.MultipleImages: %d", options.MultipleImages)
+	}
```
Broken out as `validateImageSelection`.

nalind commented on copy/copy.go:
```diff
@@ -205,80 +238,279 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
-		}
-		logrus.Debugf("Source is a manifest list; copying (only) instance %s", instanceDigest)
-		unparsedInstance := image.UnparsedInstance(rawSource, &instanceDigest)
+		if options.MultipleImages == CopyOnlyCurrentRuntimeImage {
```
Added a comment by the call to `ChooseInstance()`, which is where that selection is made.

nalind commented on copy/copy.go:
```diff
@@ -205,80 +238,279 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
+			if err != nil {
+				return nil, errors.Wrapf(err, "Error choosing an image from manifest list %s", transports.ImageName(srcRef))
+			}
+			logrus.Debugf("Source is a manifest list; copying (only) instance %s", instanceDigest)
```
Adding a `for current system` to the end of that message.

nalind commented on copy/copy.go:
```diff
@@ -205,80 +238,279 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
+		} else { /* options.MultipleImages == CopyAllImages or options.MultipleImages == CopyOnlyInstances, */
+			// If we were asked to copy multiple images and can't, that's an error.
+			if !supportsMultipleImages(c.dest) {
+				return nil, errors.Errorf("Error copying multiple images: destination reference %q does not support multiple images", transports.ImageName(destRef))
```
Changed it to just print the transport name, though for transports like `docker`, it's not a limitation of the transport so much of the registry to which we're communicating, but that's probably something the user who gets the error message is already going to know anyway.

nalind commented on copy/copy.go:
```diff
@@ -205,80 +238,279 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
+					}
+				}
 			}
 			if !matches {
```
Moved a couple of things around, but we do need to check `matches` a second time, in case its value changes after we check it the first time.

nalind commented on copy/copy.go:
```diff
@@ -205,80 +238,279 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
+				return nil, "", "", errors.Wrapf(err, "Error computing digest of source image's manifest")
+			}
+			if !matches {
+				if value := ctx.Value(internal.CopyUnparsedTopImageKey); value != nil {
```
Changed the `Commit()` API, so this is gone now.

nalind commented on copy/manifest.go:
```diff
@@ -119,3 +119,38 @@ func isMultiImage(ctx context.Context, img types.UnparsedImage) (bool, error) {
+func (c *copier) determineListConversion(currentListMIMEType string, destSupportedListMIMETypes []string, forcedListMIMEType string) (string, error) {
+	// If there's no list of supported types, then anything we support is expected to be supported.
+	if len(destSupportedListMIMETypes) == 0 {
+		destSupportedListMIMETypes = manifest.SupportedListMIMETypes
```
Yes, that behavior predates this PR.

nalind commented on copy/manifest.go:
```diff
@@ -119,3 +119,38 @@ func isMultiImage(ctx context.Context, img types.UnparsedImage) (bool, error) {
+	// The lowest priority is the first member of the list of acceptable types that is a list.
+	selectedType := destSupportedListMIMETypes[0]
+	if !manifest.MIMETypeIsMultiImage(selectedType) {
+		for i := range destSupportedListMIMETypes {
```
Reworked this.

nalind commented on copy/manifest.go:
```diff
@@ -119,3 +119,38 @@ func isMultiImage(ctx context.Context, img types.UnparsedImage) (bool, error) {
+	if len(destSupportedListMIMETypes) == 0 {
+		destSupportedListMIMETypes = manifest.SupportedListMIMETypes
+	}
+	// The lowest priority is the first member of the list of acceptable types that is a list.
```
I think it's confusing this way, so it's been reworked.

nalind commented on copy/manifest.go:
```diff
@@ -119,3 +119,38 @@ func isMultiImage(ctx context.Context, img types.UnparsedImage) (bool, error) {
+		}
+	}
+	if forcedListMIMEType != "" {
+		// If we're forcing it, we prefer the forced value over everything else.
```
This should be cleaner in the newer version.

nalind commented on copy/manifest.go:
```diff
@@ -119,3 +119,38 @@ func isMultiImage(ctx context.Context, img types.UnparsedImage) (bool, error) {
+
+// determineListConversion returns the MIME type to which we should convert a list of manifests.
+// Returns the preferred manifest MIME type (whether we are converting to it or using it unmodified),
+// and a list of other possible alternatives, in order.
```
Updated.

nalind commented on copy/manifest.go:
```diff
@@ -119,3 +119,38 @@ func isMultiImage(ctx context.Context, img types.UnparsedImage) (bool, error) {
+				selectedType = allowedType
+				break
+			}
+		}
```
Reworked it.

nalind commented on docker/docker_image_dest.go:
```diff
@@ -343,20 +345,36 @@ func (d *dockerImageDestination) TryReusingBlob(ctx context.Context, info types.
 // PutManifest writes manifest to the destination.
+// If instanceDigest is not nil, it contains a digest of the specific manifest instance to overwrite the manifest for (when
+// the primary manifest is a manifest list); this should always be nil if the primary manifest is not a manifest list.
+// It is expected but not enforced that the instanceDigest, when specified, matches the digest of `manifest` as generated
```
Went ahead and started enforcing it.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,212 @@
+// Instance returns the size and MIME type of a particular instance in the list.
+func (list *Schema2List) Instance(instanceDigest digest.Digest) (ListUpdate, error) {
+	for _, manifest := range list.Manifests {
+		if manifest.Digest == instanceDigest {
```
We don't guarantee uniqueness among entries in the list.  I don't understand how a `map[digest]int` would change that?

vrothberg replied:
Can it lead to issues that we're non-unique digests? We could think of referencing by their offset in the list/index?

nalind replied:
The list gets searched either for a particular digest, as we do here, or for an entry that has specific values in its corresponding platform fields, as in the `ChooseInstance()` implementations.

Multiple entries with the same instance digest should be referring to the same image since they would have the same manifest, and by extension the same layers and configuration, so for the values we're returning here, the only meaningful difference would be in the MediaType, which is often discarded.

Multiple entries with the same instance digests but different platform values shouldn't break anything, but I guess that's a problem for `ChooseInstance()` to worry about.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,212 @@
+	return results
+}
+
+// Instance returns the size and MIME type of a particular instance in the list.
```
Changed it.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,212 @@
+		return list.Clone(), nil
+	case imgspecv1.MediaTypeImageIndex:
+		return list.ToOCI1Index()
+	case DockerV2Schema1MediaType, DockerV2Schema1SignedMediaType, imgspecv1.MediaTypeImageManifest, DockerV2Schema2MediaType:
```
Moved the logic after the `switch` to be its `default:` clause.

nalind commented on manifest/manifest.go:
```diff
@@ -241,9 +247,11 @@ func FromBlob(manblob []byte, mt string) (Manifest, error) {
 		return nil, fmt.Errorf("Treating manifest lists as individual manifests is not implemented")
-	default: // Note that this may not be reachable, NormalizedMIMEType has a default for unknown values.
-		return nil, fmt.Errorf("Unimplemented manifest MIME type %s", mt)
+	case imgspecv1.MediaTypeImageIndex:
```
Combined them.  I guess the MIME type is known by the caller, so we don't need to hint at it in the error text.

nalind commented on manifest/oci_index.go:
```diff
@@ -0,0 +1,217 @@
+	return imgspecv1.MediaTypeImageIndex
+}
+
+// Instances returns a index of the manifests that this index knows of.
```
Changed it.

nalind commented on manifest/oci_index.go:
```diff
@@ -0,0 +1,217 @@
+	return results
+}
+
+// Instance returns the size and MIME type of a particular instance in the index.
```
Changed it.

nalind commented on manifest/oci_index.go:
```diff
@@ -0,0 +1,217 @@
+		if updates[i].MediaType == "" {
+			return errors.Errorf("update %d of %d passed to OCI1Index.UpdateInstances had no media type", i+1, len(updates))
+		}
+		index.Manifests[i].MediaType = updates[i].MediaType
```
Added calls to `SupportedOCI1MediaType()` and `SupportedSchema2MediaType()`, and a check to explicitly allow the index type, which the spec says we should allow.

nalind commented on manifest/docker_list.go:
```diff
@@ -0,0 +1,212 @@
+		if updates[i].MediaType == "" {
+			return errors.Errorf("update %d of %d passed to Schema2List.UpdateInstances had no media type", i+1, len(updates))
+		}
+		list.Manifests[i].MediaType = updates[i].MediaType
```
Added calls to `SupportedSchema2MediaType()` and `SupportedOCIMediaType()` here.  I don't have confirmation that lists are allowed to mention other lists in this format.

nalind commented:
Updated.

vrothberg commented on copy/copy.go:
```diff
@@ -121,12 +153,24 @@ type Options struct {
+		return errors.Errorf("Invalid value for options.ImageListSelection: %d", selection)
+	case CopySystemImage, CopyAllImages, CopySpecificImages:
+	}
+	return nil
```
The return in the outer scope here isn't needed. The go compiler unwraps the branches and won't complain if all have a return path. So we could rewrite the function body into:
```golang
	switch selection {
	case CopySystemImage, CopyAllImages, CopySpecificImages:
		return nil
	default:
		return errors.Errorf("Invalid value for options.ImageListSelection: %d", selection)
	}
```

My brain-parser exited here since its used to the default being the last case.

nalind replied:
Reordered to `return nil` from the known-values cases, and to put the `default:` clause last.

vrothberg commented on copy/copy.go:
```diff
@@ -205,80 +238,279 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
+	// Copy each image, or just the ones we want to copy, in turn.
+	instanceDigests := list.Instances()
+	updates := make([]manifest.ListUpdate, len(instanceDigests))
+	for i, instanceDigest := range instanceDigests {
```
This is still valid but I am cool to defer beautifying the copy reports to a future PR.

nalind replied:
For whatever reason I read this as needing to be done using a progress bar, but now that I look at it again, it sounds much easier.

nalind replied:
Okay, non-debug text added.

vrothberg commented on copy/manifest.go:
```diff
@@ -119,3 +119,36 @@ func isMultiImage(ctx context.Context, img types.UnparsedImage) (bool, error) {
+	}
+	if forcedListMIMEType != "" {
+		// If we're forcing it, we prefer the forced value over everything else.
+		selectedType = forcedListMIMEType
```
I think we can move this branch to the top and just return `forcedListMIMEType`. Although it looks a bit complicated to return an argument (if it's set).

nalind replied:
Moved it.

vrothberg commented on docker/docker_image_dest.go:
```diff
@@ -343,20 +345,43 @@ func (d *dockerImageDestination) TryReusingBlob(ctx context.Context, info types.
+			return errors.Wrapf(err, "error digesting manifest in PutManifest")
+		}
+		if !matches {
+			return errors.New("Attempted to PutManifest using an explicitly specified digest that didn't match the manifest")
```
Can we add the two digests to the error?

nalind replied:
Sure can!  Added.

nalind commented:
Dropped the patch that redirected to https://github.com/containers/skopeo/pull/741 for testing, so I expect the `test-skopeo` target to fail.

rhatdan commented:
LGTM
As expected without the DONOTMERGE patch, tests fail.  
Merging.

grooverdan commented:
Thank you all for writing and reviewing this feature. :cake:
